modules/calib3d/src/ippe.cpp

```diff
@@ -0,0 +1,1124 @@
+        b2 = -imgPoints.at<Vec2d>(i)(1);
+
+        ATA02 = ATA02 + a2;
+        ATA12 = ATA12 + b2;
+        ATA20 = ATA20 + a2;
+        ATA21 = ATA21 + b2;
+        ATA22 = ATA22 + a2 * a2 + b2 * b2;
+
+        bx = -a2 * rz - rx;
+        by = -b2 * rz - ry;
+
+        ATb0 = ATb0 + bx;
+        ATb1 = ATb1 + by;
+        ATb2 = ATb2 + a2 * bx + b2 * by;
+    }
+
+    double detAInv = 1.0 / (ATA00 * ATA11 * ATA22 - ATA00 * ATA12 * ATA21 - ATA02 * ATA11 * ATA20);
+
+    //construct S:
+    S00 = ATA11 * ATA22 - ATA12 * ATA21;
+    S01 = ATA02 * ATA21;
+    S02 = -ATA02 * ATA11;
+    S10 = ATA12 * ATA20;
+    S11 = ATA00 * ATA22 - ATA02 * ATA20;
+    S12 = -ATA00 * ATA12;
+    S20 = -ATA11 * ATA20;
+    S21 = -ATA00 * ATA21;
+    S22 = ATA00 * ATA11;
+
+    //solve t:
+    Mat t = _t.getMat();
+    t.at<double>(0) = detAInv * (S00 * ATb0 + S01 * ATb1 + S02 * ATb2);
+    t.at<double>(1) = detAInv * (S10 * ATb0 + S11 * ATb1 + S12 * ATb2);
+    t.at<double>(2) = detAInv * (S20 * ATb0 + S21 * ATb1 + S22 * ATb2);
+}
+
+void PoseSolver::computeRotations(double j00, double j01, double j10, double j11, double p, double q, OutputArray _R1, OutputArray _R2)
+{
+    //This is fairly optimized code which makes it hard to understand. The matlab code is certainly easier to read.
+    _R1.create(3, 3, CV_64FC1);
+    _R2.create(3, 3, CV_64FC1);
+
+    double a00, a01, a10, a11, ata00, ata01, ata11, b00, b01, b10, b11, binv00, binv01, binv10, binv11;
+    //double rv00, rv01, rv02, rv10, rv11, rv12, rv20, rv21, rv22;
+    double rtilde00, rtilde01, rtilde10, rtilde11;
+    double rtilde00_2, rtilde01_2, rtilde10_2, rtilde11_2;
+    double b0, b1, gamma, dtinv;
+    double sp;
+
+    Mat Rv;
+    Mat v(3,1,CV_64FC1);
+    v.at<double>(0) = p;
+    v.at<double>(1) = q;
+    v.at<double>(2) = 1;
+    rotateVec2ZAxis(v,Rv);
+    Rv = Rv.t();
+
+    //setup the 2x2 SVD decomposition:
+    double rv00, rv01, rv02;
+    double rv10, rv11, rv12;
+    double rv20, rv21, rv22;
+    rv00 = Rv.at<double>(0,0);
+    rv01 = Rv.at<double>(0,1);
+    rv02 = Rv.at<double>(0,2);
+
+    rv10 = Rv.at<double>(1,0);
+    rv11 = Rv.at<double>(1,1);
+    rv12 = Rv.at<double>(1,2);
+
+    rv20 = Rv.at<double>(2,0);
+    rv21 = Rv.at<double>(2,1);
+    rv22 = Rv.at<double>(2,2);
+
+    b00 = rv00 - p * rv20;
+    b01 = rv01 - p * rv21;
+    b10 = rv10 - q * rv20;
+    b11 = rv11 - q * rv21;
+
+    dtinv = 1.0 / ((b00 * b11 - b01 * b10));
+
+    binv00 = dtinv * b11;
+    binv01 = -dtinv * b01;
+    binv10 = -dtinv * b10;
+    binv11 = dtinv * b00;
+
+    a00 = binv00 * j00 + binv01 * j10;
+    a01 = binv00 * j01 + binv01 * j11;
+    a10 = binv10 * j00 + binv11 * j10;
+    a11 = binv10 * j01 + binv11 * j11;
+
+    //compute the largest singular value of A:
+    ata00 = a00 * a00 + a01 * a01;
+    ata01 = a00 * a10 + a01 * a11;
+    ata11 = a10 * a10 + a11 * a11;
+
+    gamma = sqrt(0.5 * (ata00 + ata11 + sqrt((ata00 - ata11) * (ata00 - ata11) + 4.0 * ata01 * ata01)));
```

Comment:
Should we check?
- `sqrt()` argument is not negative (due rounding errors)
- `gamma` is not close to zero (division is below) or raise `CV_Error(Error::StsNoConv, "...");`

Reply:
I have added the two checks with `CV_Error`.